pkg/providers/v1/zones.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"k8s.io/klog/v2"
+)
+
+type zoneDetails struct {
+	name     string
+	id       string
+	zoneType string
+}
+
+type zoneCache struct {
+	cloud             *Cloud
+	mutex             sync.Mutex
+	zoneNameToDetails map[string]zoneDetails
+}
+
+// Get the zone details by zone names and load from the cache if available as
+// zone information should never change.
+func (z *zoneCache) getZoneDetailsByNames(zoneNames []string) (map[string]zoneDetails, error) {
+	if len(zoneNames) == 0 {
+		return map[string]zoneDetails{}, nil
+	}
+
+	if z.shouldPopulateCache(zoneNames) {
+		// Populate the cache if it hasn't been populated yet
+		err := z.populate()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	requestedZoneDetails := map[string]zoneDetails{}
+	for _, zone := range zoneNames {
+		if zoneDetails, ok := z.zoneNameToDetails[zone]; ok {
+			requestedZoneDetails[zone] = zoneDetails
+		} else {
+			klog.Warningf("Could not find zone %s", zone)
+		}
+	}
```

Comment:
Not super sure but I think we need to `RLock` around here and in `shouldPopulateCache` :)

Reply:
Good catch. From reading stack overflow, sounds like reading a map is only thread safe if that map doesn't get written to.

Reply:
I've just moved the mutex to the top-level function as that's where all of the other functions are invoked. I added a note on the functions that they're not thread safe if not called from here, and I don't know the cleanest way to be able to compose this into multiple functions while ensuring that someone doesn't use them in an unsafe way. I left comments for now.